### PR TITLE
Add cross-platform exception serialization support

### LINKED ISSUE #7215: System messages containing BCL exceptions fail to deserialize on netfx when serialized on net5+/.net core
**Version Information**
Akka.Remote v1.5.20

**Describe the bug**
I'm currently running a mixed net8.0/net481 Akka cluster.  All the following proto messages contain an `ExceptionData` field:

* Akka.Remote.Serialization.Proto.Msg.StatusFailure
* Akka.Remote.Serialization.Proto.Msg.RecreateData
* Akka.Remote.Serialization.Proto.Msg.ResumeData
* Akka.Remote.Serialization.Proto.Msg.FailedData

 When one of the above messages containing a BCL exception is serialized on a net8.0 node, sent, and then deserialized on a net481 node, I get the following deserialization exception.

 ```csharp
Value cannot be null. Parameter name: type
at System.Runtime.Serialization.SerializationInfo..ctor(Type type, IFormatterConverter converter, Boolean requireSameTokenInPartialTrust)
at Akka.Remote.Serialization.ExceptionSupport.ExceptionFromProtoNet(ExceptionData proto)
at Akka.Remote.Serialization.SystemMessageSerializer.RecreateFromProto(Byte[] bytes)
at Akka.Serialization.Serialization.Deserialize(Byte[] bytes, Int32 serializerId, String manifest)
at Akka.Remote.MessageSerializer.Deserialize(ExtendedActorSystem system, Payload messageProtocol)
at Akka.Remote.DefaultMessageDispatcher.Dispatch(IInternalActorRef recipient, Address recipientAddress, Payload message, IActorRef senderOption)
at Akka.Remote.EndpointReader.b__16_0(Message msg)
at Akka.Util.Internal.Extensions.ForEach[T](IEnumerable1 source, Action1 action)
at Akka.Remote.EndpointReader.b__11_0(InboundPayload inbound)
at lambda_method(Closure , Object , Action1 , Action1 , Action1 ) at Akka.Tools.MatchHandler.PartialHandlerArgumentsCapture4.Handle(T value)
at Akka.Actor.ReceiveActor.ExecutePartialMessageHandler(Object message, PartialAction`1 partialAction)
at Akka.Actor.UntypedActor.Receive(Object message)
at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
at Akka.Actor.ActorCell.ReceiveMessage(Object message)
at Akka.Actor.ActorCell.Invoke(Envelope envelope)]
```
The issue is because in netfx BCL exceptions are defined in the mscorlib assembly, and in net5+/net core they are defined in the System.Private.Corlib assembly.  The issue can be resolved if we are a bit smarter about how we define the proto message `ExceptionData.TypeName` field for these serialized payloads.  If we are going to use the type name as the manifest, we should respect the `TypeForwardedFromAttribute` if one exists.  This is exactly what the binary formatter does (see [BinaryFormatter.GetTypeInformation()](https://source.dot.net/#System.Runtime.Serialization.Formatters/System/Runtime/Serialization/Formatters/Binary/BinaryFormatter.cs,300096fe7ef6f757)).

Additional details and info can also be found in the discussion: https://github.com/akkadotnet/akka.net/discussions/7193

**To Reproduce**
https://github.com/BrettJaner/akka-exception-serialization-issue

**Expected behavior**
No deserialization exception on the netfx nodes.

**Actual behavior**
Deserialization exceptions on the netfx nodes.

**Environment**
Windows

**Potential Fix**
***Update the following extension method from***
```csharp
namespace Akka.Util
{
    public static class TypeExtensions
    {
        public static string TypeQualifiedName(this Type type)
        {
            if (ShortenedTypeNames.TryGetValue(type, out var shortened))
            {
                return shortened;
            }

            shortened = cleanAssemblyVersionRegex.Replace(type.AssemblyQualifiedName, string.Empty);
            ShortenedTypeNames.TryAdd(type, shortened);
            
            return shortened;
        }
    }
}
```
***To***
```csharp
namespace Akka.Util
{
    public static class TypeExtensions
    {
        public static string TypeQualifiedName(this Type type)
        {
            if (ShortenedTypeNames.TryGetValue(type, out var shortened))
            {
                return shortened;
            }

            // respect type forwarded from attribute since types can move assemblies between .net versions
            var typeForwardedFromAttribute = type.GetCustomAttribute<TypeForwardedFromAttribute>();

            if (typeForwardedFromAttribute != null)
            {
                var assemblyName = cleanAssemblyVersionRegex.Replace(typeForwardedFromAttribute.AssemblyFullName, string.Empty);

                shortened = $"{type.FullName}, {assemblyName}";
            }
            else
            {
                shortened = cleanAssemblyVersionRegex.Replace(type.AssemblyQualifiedName, string.Empty);
            }

            ShortenedTypeNames.TryAdd(type, shortened);
            
            return shortened;
        }
    }
}
```

### PR DESCRIPTION
**Pull request by Arkatufus**
Fixes #7215

## Changes

Embed `TypeForwardedFromAttribute` data inside the wire format so we can map `System.Private.Corlib` to `mscorlib` types.

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).

### CODE REVIEW

File: src/core/Akka.Remote/Serialization/ExceptionSupport.cs [hunk 1]
```diff
@@ -82,6 +83,11 @@ public Proto.Msg.ExceptionData ExceptionToProtoNet(Exception exception)
             message.StackTrace = exception.StackTrace ?? "";
             message.Source = exception.Source ?? "";
             message.InnerException = ExceptionToProto(exception.InnerException);
+            
+            var forwardedFrom = exceptionType.GetCustomAttribute<TypeForwardedFromAttribute>();
```
Comment (Arkatufus) [inline src/core/Akka.Remote/Serialization/ExceptionSupport.cs:87]: .NET standard introduced this attribute to mark a type as being moved from a different assembly

File: src/core/Akka.Remote/Serialization/ExceptionSupport.cs [hunk 2]
```diff
@@ -110,7 +116,19 @@ public Exception ExceptionFromProtoNet(Proto.Msg.ExceptionData proto)
...
-            Type exceptionType = Type.GetType(proto.TypeName);
+            var exceptionType = Type.GetType(proto.TypeName);
+            
+            // If type loading failed and type was forwarded from an older assembly,
+            // retry by loading the type from the older assembly name
+            if (exceptionType is null && proto.TypeForwardedFrom != string.Empty)
+            {
+                var typeName = $"{proto.TypeName[..proto.TypeName.IndexOf(',')]}, {proto.TypeForwardedFrom}";
+                exceptionType = Type.GetType(typeName);
+            }
```
Comment (Arkatufus) [inline src/core/Akka.Remote/Serialization/ExceptionSupport.cs:127]: Retry type loading with a different assembly name, if the type was marked as moved from an older assembly name

File: src/protobuf/ContainerFormats.proto [hunk 3]
```diff
@@ -69,6 +69,7 @@ message ExceptionData {
   string source = 4;
   ExceptionData innerException = 5;
   map<string, Payload> customFields = 6;
+  string typeForwardedFrom = 7;
```
Comment (Arkatufus) [inline src/protobuf/ContainerFormats.proto:72]: Added a single entry in the .proto file

**Review by Arkatufus — COMMENTED** [review]
Self review

**Comment by Arkatufus** [comment]
Limiting the blast radius of this change to remoting `ExceptionData` (de)serialization, we're not applying this technique to all of our serialization pipeline

File: src/core/Akka.Remote/Serialization/ExceptionSupport.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/core/Akka.Remote/Serialization/ExceptionSupport.cs:87]: Wow, that would have been good to know a long time ago lol

File: src/core/Akka.Remote/Serialization/ExceptionSupport.cs (hunk 2 above)
Comment (Aaronontheweb) [reply src/core/Akka.Remote/Serialization/ExceptionSupport.cs:127]: LGTM

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Comment by Aaronontheweb** [comment]
Discussed with @Arkatufus - we're gonna add a repro to this before we merge it.

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**